### ISSUE TITLE
[Bug] NaN's if a single observation is sampled from PPO's rollout_buffer

### ISSUE THREAD

**Issue by dHonerkamp** [DLR-RM/stable-baselines3#325]
### 🐛 Bug

I am using PPO and am adapting a few things of it to experiment with an unusual adversary setting, including for now in an ad-hoc manner adjusting the rollout-buffer way to have a dynamic size for it. Doing this I found that PPO can run into nan values whenever it samples a batch size of size 1 due to this in `PPO.py`, line 170:

```
                advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
```
where `advantages.std()` over a batch of size 1 is not defined.

While my current scenario is of course nothing common, this seems to me like a case that could more generally occur whenever someone uses a btach_size of size rollout_buffer.size + 1 or someone wants to use a dynamically sized replay buffer. It could also easily be accounted for.

### To Reproduce

Run PPO with a batch_size of 1 or a rollout_buffer of size batch_size + 1 (i.e. by setting `n_steps` to this value)

### Expected behavior

I would suggest to either discard batches when only a single observation is left in the rollout_buffer or check the size of the sampled buffer and set the std to either 1 or the minimum value in this case.

### System Info

stable-baselines v. 0.10.0

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [X] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by araffin** [comment]
> I have checked that there is no similar issue in the repo (required)

Duplicate of https://github.com/DLR-RM/stable-baselines3/issues/265

if you want to use batch_size=1 anyway, you need to specify `unbiased=False` when computing the std (cf. PyTorch doc).

**Comment by externalsupplierstaff** [comment]
I see
Why don't `thing.std(unbiased=len(thing)>1)`, or a warning, or something?
This is a silent and hard to track bug, it's not nice people keep stumbling upon it.

**Comment by hughperkins** [comment]
I encountered this issue too. I spent 4 days wondering why my training was naning out. Eventually started to debug, and discovered it was because my batch_size was not a factor of `n_env * n_step`. I have my own reasons why this is not the case. It is not an accident. I did see the warning that batch_size is recommended to be a factor of `n_env * n_step`, since last batch will be truncated. However I did not expect my training to fail with nans. Basically, for PPO, if batch size is not a factor, then each learning stage, there is a probability of 1: batch_size of the buffer returning batch size of 1 for last batch, and training failing with nan.

I intend to propose a PR to address this issue. Two possible solutions:
1. PPO can reject batches of either size 1, or less than batch_size
2. rollout buffer can yield only full batches

For now I'm going with option 2, since easy, and I don't see a need to yield non-full batches. However, I do see the advantage of solution 1, and can be persuaded to choose solution 1 instead. (However, I think that most people have batch_size be a factor of n_env * n_step, and so this change won't affect them, no matter which option I choose).

**Comment by araffin** [comment]
Hello,
i would be in favor of https://github.com/DLR-RM/stable-baselines3/issues/325#issuecomment-843277873 (as it is only due to advantage normalization)

you also have a 4th option which is disabling advantage normalization

**Comment by hughperkins** [comment]
Created https://github.com/DLR-RM/stable-baselines3/pull/1025

**Comment by hughperkins** [comment]
(didn't see your comment whilst working on the PR).

**Comment by hughperkins** [comment]
I've used up all my energy on writing tests and so on for approach 2. Will leave for now, and perhaps come back to modify how advantage works. But... do we want to train on partial batches? The learning rate will be wrong for the partial batches, and the gradient will be all over the place. Can you think of any occasion where we would want to train on partial batches?

**Comment by araffin** [comment]
> (didn't see your comment whilst working on the PR).

that's why we recommend to discuss and agree on the solution first in an issue... I will still try to take a look at your PR in the coming days.

Thinking about it again, I think the best would be to skip advantage normalization if the batch size is of size 1.

>  But... do we want to train on partial batches?

we usually want to train on all the collected data.

> The learning rate will be wrong for the partial batches,

that's true, at the same time there should be only one partial batch.

>  if batch size is not a factor, then each learning stage, there is a probability of 1: batch_size of the buffer returning batch size of 1 for last batch

this should not be a probability. It return a batch of size 1 if `buffer_size % batch_size == 1`.
For instance:
```python
from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env

# With one env
n_envs = 1
n_steps = 64
batch_size = 63
# With two
# n_envs = 2
# n_steps = 32
# batch_size = 63

buffer_size = n_envs * n_steps
assert buffer_size % batch_size == 1

env = make_vec_env("Pendulum-v1", n_envs=n_envs)
# Fix1: normalize_advantage=False
PPO("MlpPolicy", env, verbose=1, batch_size=batch_size, n_steps=n_steps).learn(10_000)
```

**Comment by hughperkins** [comment]
My buffer size is not fixed. I collect the roll out myself, using a custom
method, therefore it is a probability. It fails one time in batch size. I
bumped the batch size up to 2048, and the nans were less frequent. I
thought this was because batch size 2048 stabilized the learning. But it
was because the bug only occurs with probability 1/batch size, for me

**Comment by hughperkins** [comment]
I think there are two issues we can factorize:
- how to avoid the nans => the advantage fix you allude to. I agree
- should partial batches be skipped => this would be backwards
incomptaible, so should be a new option, that defaults to not skipping
partial batches.

For now, I'm just going to leave this code and comments here. Maybe I will
submit a PR for either or both of these. My own code works now anyway, and
I've made the code available that I'm using, via this PR, if anyone else
wants to use it.

**Comment by hughperkins** [comment]
Since this is an actual bug, that is REALLY PAINFUL to encounter the first time, took me 40 hours to solve, could we please leave this open in the meantime?

**Comment by hughperkins** [comment]
(I mean, basically it never occurred to me to search in closed issues until
I'd already created a PR and started to submit it. Don't suppose... Could
we open this issue please? It might save other people a lot of pain, and
someone might fix it.)

**Comment by araffin** [comment]
@hughperkins not sure to understand your comment... this issue was fixed by yourself in #1028

**Comment by hughperkins** [comment]
That's a very old email. Not sure why it is appearing as a comment now...

**Comment by hughperkins** [comment]
Oh, maybe emails are rate limited, so you will receive these new emails in
December and February respectively, lol :D

**Comment by hughperkins** [comment]
I sent that email in August :P
